### PR TITLE
Support for running xfstests on loop device

### PR DESCRIPTION
**Pull request by harish-24**
1.Enables tests to be run on loop device
2.Runs all tests for given file system
(or)
Runs specific tests for given file system if range is provided

Signed-off-by: Harish <harish@linux.vnet.ibm.com>

### CODE REVIEW

File: fs/xfstests.py [hunk 1]
```diff
@@ -84,42 +96,90 @@ def setUp(self):
                       ", ".join(self.available_tests))
         process.run('useradd fsgqa', sudo=True)
         process.run('useradd 123456-fsgqa', sudo=True)
+        self.scratch_mnt = '/mnt/scratch'
+        self.test_mnt = '/mnt/test'
```
Comment (nasastry) [inline fs/xfstests.py]: If possible make the above two `scratch_mnt, test_mnt` as configurable.

File: fs/xfstests.py [hunk 2]
```diff
@@ -84,42 +96,90 @@ def setUp(self):
...
+            os.rmdir(self.test_mnt)
+        if self.dev_type == 'loop':
+            process.system(
+                'for i in $(seq 0 1); do losetup -d '
+                '/dev/loop$i; done', shell=True, sudo=True)
+        self.part.unmount()
+
+    def _create_loop_device(self, base_disk):
+        self.part = partition.Partition(
+            base_disk, mountpoint='/mnt/loop-device')
```
Comment (nasastry) [inline fs/xfstests.py]: If possible make the above `/mnt/loop-device` as configurable.

File: fs/xfstests.py [hunk 3]
```diff
@@ -84,42 +96,90 @@ def setUp(self):
...
+            'for i in $(seq 0 1); do fallocate -o 0 -l 9GiB '
+            '/mnt/loop-device/file-$i.img; done', shell=True, sudo=True)
+        process.run(
+            'for i in $(seq 0 1); do losetup /dev/loop$i '
+            '/mnt/loop-device/file-$i.img; done', shell=True, sudo=True)
+        if process.system('which mkfs.%s' % self.fs_to_test):
+            self.skip('Unknown filesystem %s' % self.fs_to_test)
+        # mkfs for two loop device
+        for i in range(2):
+            loop_dev = partition.Partition('/dev/loop%s' % i)
```
Comment (nasastry) [inline fs/xfstests.py]: What happens if the `/dev/loop1 or /dev/loop2` is busy?

File: fs/xfstests.py [hunk 4]
```diff
@@ -43,17 +43,20 @@ def setUp(self):
...
 
         elif detected_distro.name in ['centos', 'fedora', 'redhat']:
             packages.extend(['acl', 'bc', 'dump', 'indent', 'libtool', 'lvm2',
-                             'xfsdump', 'psmisc', 'sed', 'libacl-devel', 'libattr-devel',
-                             'libaio-devel', 'libuuid-devel', 'openssl-devel', 'xfsprogs-devel',
-                             'btrfs-progs-devel'])
+                             'xfsdump', 'psmisc', 'sed', 'libacl-devel',
+                             'libattr-devel', 'libaio-devel', 'libuuid-devel',
+                             'openssl-devel', 'xfsprogs-devel', 'btrfs-progs-devel'])
+            if detected_distro.name != 'redhat':
```
Comment (nasastry) [inline fs/xfstests.py]: Wrong indentation here. This `if` condition is under `Ubuntu`. I guess you wanted `redhat` condition parallel to `Ubuntu` not under `Ubuntu`.

File: fs/xfstests.py [hunk 5]
```diff
@@ -84,42 +96,90 @@ def setUp(self):
...
+                    self.log.info('ERR: %s failed. Message: %s', test, msg)
+                    failures = True
         if failures:
             self.fail('One or more tests failed. Please check the logs.')
 
     def tearDown(self):
         process.system('userdel fsgqa', sudo=True)
         process.system('userdel 123456-fsgqa', sudo=True)
+        # In case if any test has been interrupted
+        process.system('umount self.scratch_mnt self.test_mnt',
```
Comment (nasastry) [inline fs/xfstests.py]: The above `umount self.scratch_mnt self.test_mnt` should be like
`process.system('umount %s %s' % (self.scratch_mnt, self.test_mnt), sudo=True, ignore_status=True)`

File: fs/xfstests.py [hunk 6]
```diff
@@ -84,42 +96,90 @@ def setUp(self):
...
+                       sudo=True, ignore_status=True)
+        if os.path.exists(self.scratch_mnt):
+            os.rmdir(self.scratch_mnt)
+        if os.path.exists(self.test_mnt):
+            os.rmdir(self.test_mnt)
+        if self.dev_type == 'loop':
+            process.system(
+                'for i in $(seq 0 1); do losetup -d '
+                '/dev/loop$i; done', shell=True, sudo=True)
```
Comment (nasastry) [inline fs/xfstests.py]: How are you doing handle error condition over here (if `/dev/loop0, /dev/loop1` is not availble)

File: fs/xfstests.py [hunk 7]
```diff
@@ -62,10 +65,19 @@ def setUp(self):
...
+            base_disk = self.params.get('base_disk', default=None)
+            if not base_disk:
+                self.skip('Please provide a base disk to create loop device')
+            self._create_loop_device(base_disk)
+        else:
+            self.test_dev = self.params.get('test_dev', default=None)
+            self.scratch_dev = self.params.get('scratch_dev', default=None)
+            # TODO: Overwirte the disk in local.config
 
         git.get_repo('git://git.kernel.org/pub/scm/fs/xfs/xfstests-dev.git',
```
Comment (nasastry) [inline fs/xfstests.py]: Can this be placed in a yaml file?

Comment (PraveenPenguin) [inline fs/xfstests.py]: i think this should be ok as it not going to change in future

File: fs/xfstests.py.data/xfstests.yaml [hunk 8]
```diff
@@ -1,4 +1,16 @@
...
+    skip_dangerous: True
+    file_system: 'ext4'
+
+    # Uncomment and edit test_range for running specific tests
+    # test_range: '73,217-415'
+
+    # Run with either loop_type (or) disk_type
+    loop_type: !mux
+        type: 'loop'
+        base_disk: /dev/sdj2
```
Comment (narasimhan-v) [inline fs/xfstests.py.data/xfstests.yaml]: Can this parameter name be `disk` please, as it is uniform in all other tests

File: fs/xfstests.py (hunk 4 above)
Comment (harish-24) [inline fs/xfstests.py]: No. I have placed  `if` inside `['centos', 'fedora', 'redhat']` case and handled separately for `redhat` as most of packages in this group are similar.

File: fs/xfstests.py (hunk 1 above)
Comment (harish-24) [inline fs/xfstests.py]: sure

File: fs/xfstests.py (hunk 5 above)
Comment (harish-24) [inline fs/xfstests.py]: yes, thanks, will replace that.

File: fs/xfstests.py (hunk 2 above)
Comment (harish-24) [inline fs/xfstests.py]: sure

File: fs/xfstests.py.data/xfstests.yaml (hunk 8 above)
Comment (harish-24) [inline fs/xfstests.py.data/xfstests.yaml]: @narasimhan-v @PraveenPenguin It is okay to change. I have added a separate type for direct disks, currently in `TODO`

File: fs/xfstests.py (hunk 4 above)
Comment (nasastry) [inline fs/xfstests.py]: Yes, you are right. Because of big line I missed it.

File: fs/xfstests.py (hunk 3 above)
Comment (harish-24) [inline fs/xfstests.py]: I think the losetup cmd would have failed in that case.

File: fs/xfstests.py.data/xfstests.yaml (hunk 8 above)
Comment (PraveenPenguin) [inline fs/xfstests.py.data/xfstests.yaml]: sure

**Comment by harish-24** [comment]
Closing this. Opened #341

**Pull request closed without merging**